cli/src/diff_util.rs

```diff
@@ -387,13 +387,20 @@ fn show_color_words_diff_line(
                 formatter.write_all(data)?;
             }
             DiffHunk::Different(data) => {
-                let before = data[0];
-                let after = data[1];
-                if !before.is_empty() {
-                    formatter.with_label("removed", |formatter| formatter.write_all(before))?;
-                }
-                if !after.is_empty() {
-                    formatter.with_label("added", |formatter| formatter.write_all(after))?;
+                let labels = ["removed", "added"];
+                for (index, label) in labels.iter().enumerate() {
+                    let hunk_part = data[index];
+                    if !hunk_part.is_empty() {
+                        formatter.with_label(label, |formatter| {
+                            if hunk_part.iter().all(u8::is_ascii_whitespace) {
```

Comment:
I was initially thinking that we should color all added and removed whitespace but that would make common diffs that add a few lines harder to read with leading green whitespace. Without coloring all whitespace, we still risk confusing users e.g. when they replace `.` by `# ` but what you have here seems like a pretty good compromise.

Reply:
Yeah, being able to use a background color would be the best, but users can do it themselves (figure out a background color setting that works with their terminal and set it both on "diff removed" and "diff removed whitespace"). I think what's in the PR is a good compromise for a default.

Reply:
> Do you have an example where that would help?
I can't reply to this comment directly from the UI for some reason, so answering it here. What do you mean by "that"? If you mean the "all" invocation, I need that to check if hunk_part is fully whitespace.

Reply:
I think that's supposed to be https://github.com/martinvonz/jj/pull/3914#discussion_r1645409025 . I don't know why it appears as a separate comment as well. Maybe I accidentally duplicated it somehow. Anyway, I'm pretty sure the duplicate can be ignored.

Reply:
> I was initially thinking that we should color all added and removed whitespace but that would make common diffs that add a few lines harder to read with leading green whitespace.

FWIW, something like this also works if the terminal supports 256 colors and is white-on-black.
```toml
[colors]
"diff removed" = { fg = "red", bg = "#220000" }
"diff added" = { fg = "green", bg = "#002222" }
```
(I actually like it more than whitespace-only highlights because it's easier to read single-char changes.)

Reply:
I think `jj` should support the normal 16 colors. On the other hand, we could have a 256-color colorscheme as well, either as a config snippet, or as a new notion of "color schemes".

For a notion of "color schemes", we could have a color specifier for `256color diff removed` and then a way to label everything with `256color`.

Reply:
> For a notion of "color schemes",

See also https://github.com/martinvonz/jj/issues/3594

BTW, my point is, if customized bg color works, we won't need "whitespace" rule which looks a bit unstable.

Reply:
I think something like Yuya's suggestion *might* be doable with 16 colors. It's not trivial, but https://github.com/dandavison/delta 's `ansi` color scheme might do it. Here's what it looks like (note that it's different for whitespace):

![image](https://github.com/martinvonz/jj/assets/4123047/32eeede3-bb1c-44dd-985e-726a015224da)

I also checked difftastic, but it seems unable to show whitespace changes.
-----------

I think that whatever we do, it should work in 16 colors, and I don't know if that's doable without some sort of whitespace rule. If whitespace highlighting is annoying, we could only highlight changes to trailing whitespace. This is what Git does:

![image](https://github.com/martinvonz/jj/assets/4123047/02260425-c1be-48e2-80af-2d291b6c89ea)

A higher-color theme that treats the whitespace the same as everything else would be cool, but I think it'd have to be a customization.

For me, there are two reasons I like programs that use 16 colors:

- No need to worry about terminal support, even over SSH+tmux or mosh+tmux or whatever
- I picked a terminal color scheme with 16 colors I like, and non-full-screen programs using absolute color values stand out. Even if I didn't pick it out, different terminals have notably different default color schemes (xterm vs Gnome Terminal vs Alacritty are all quite different).

Reply:
As a counterpoint, I should also mention that the `ansi` color scheme is *not* the default for `delta`; I had to configure it to use it. If I were making `delta`, it'd be the default. Delta's default theme looks much too harsh to me:

![image](https://github.com/martinvonz/jj/assets/4123047/7555f8bc-8b86-4ae0-a4dc-cdae9afb01ad)

Reply:
For the purposes of this PR, should we resolve this discussion before it's submitted or can this be discussed after it's submitted? From my perspective, my main worry is introducing a new color config ("whitespace") that we will eventually deprecate if we ever decide to just change the background color of all diffs, but it looks like that will not happen (i.e. we will need the separate "whitespace" config) because we want to keep supporting 16-color terminals (which I agree with) and even if we didn't care about 16-color terminals (and also assumed that all the terminals that we care about support '#rrggbb'), there is no good default background color that will work on both dark and light backgrounds. So since we don't have that worry, it's fine to merge this first and fine-tune the display (whether through the default_refinement() mentioned earlier or something else) later.

Reply:
I think the main concern is the output looks a bit arbitrary, and stricter tokenization would make things worse? I tried this patch and it looked better than the current default, but after all, I changed the color table to highlight all removed/added regions, which produces more consistent/readable output.

I agree that there's no reasonable default of the bg color. We'll have to use underline if whitespace must be visible by default (without introducing the whitespace rule.)

Reply:
I also ended up relying on 24-bit color for myself (btw, I sent #3955 to help with that and underlining).

Actually, underlining doesn't look terrible. @jonathantanmy, @Damianrogers51, @spectral54, @ilyagr: what do you think about defaulting to underlining and then letting people override the style to use faint red/green background if their terminal supports 24-bit color?

I.e. using this as default (with #3955):
```
colors."diff removed" = { fg = "red", underline = true }
colors."diff added" = { fg = "green", underline = true }
colors."diff line_number" = { underline = false }
```

(I suspect we can assume that Googlers have 24-bit-capable terminals so we can default to using background color there.)

Reply:
Just for reference, this is the idea I mentioned in the other PR that highlights only smaller hunks (including whitespaces.)
https://github.com/martinvonz/jj/compare/main...yuja:jj:push-kzmkxxrxmoxo

Reply:
Makes sense. I just wonder if people would want to highlight whole-line diffs that consist of only whitespace. There's also the issue of trailing whitespace.

Reply:
> I.e. using this as default (with #3955)....

I don't have a definite opinion at the moment, but here's an illustration to help people think about it:

<img width="631" alt="image" src="https://github.com/martinvonz/jj/assets/4123047/d8d7cbab-dfe4-43ba-8148-c2905bfb3a6e">

Aside: The now-visible fact that it considers the spaces between `write` part of the inserted hunk was a bit distracting for me, but it makes sense if it prefers longer hunks.

Reply:
> I just wonder if people would want to highlight whole-line diffs that consist of only whitespace. There's also the issue of trailing whitespace.

I think trailing whitespaces (tabs, and some control characters?) can be labeled separately. They're usually junk, so it make sense to highlight with very bold color.

Reply:
That's fair. I'm actually torn on trailing whitespace. I don't really think it's something the VCS should help detect because it's not something people should spend time on. IMO, people who care about trailing whitespace (and others too!) should configure a formatter. But I suspect we'll get requests for it, so maybe we'll need to make it easy to style trailing whitespace differently.

Reply:
I put the color config change into a PR: https://github.com/martinvonz/jj/pull/3960

I'm fine with that too, although in my terminal I wonder if people would think it looks like underscores. I don't think it's a big concern though.

Reply:
With that PR submitted, I think that supersedes this one, so I'll close it.

Reply:
I think it's good for any diff tool to be able to show trailing whitespace, whether it's a part of a VCS or not. Same for lines containing only whitespace.

Martin's point is good 99% of the time, but I can think of at least two cases where manual formatting shows up and trailing spaces can appear:

 - Regions of source code that are excluded from auto-formatting, say Rust macros or new syntax features.
 - ASCII art or tables (on their own or inside source code comments)

Another case that we might want to consider entirely separately are changing a file from Unix to Windows line endings, or worse -- files that are a mix of these.